components/places/src/bookmark_sync/store.rs

```diff
@@ -0,0 +1,1233 @@
+                    VALUES {}",
+                    sql_support::repeat_display(chunk.len(), ",", |index, f| {
+                        let d = &chunk[index];
+                        write!(f, "(?, {}, {})", d.local_level, d.should_upload_tombstone)
+                    })
+                ),
+                chunk.iter().map(|d| d.guid.as_str()),
+            )?;
+            Ok(())
+        })?;
+
+        // "Deleting" from `itemsToMerge` fires the `insertNewLocalItems` and
+        // `updateExistingLocalItems` triggers.
+        self.db.execute_batch("DELETE FROM itemsToMerge")?;
+
+        // "Deleting" from `structureToMerge` fires the `updateLocalStructure`
+        // trigger.
+        self.db.execute_batch("DELETE FROM structureToMerge")?;
+
+        self.db.execute_batch("DELETE FROM itemsToRemove")?;
+
+        self.db.execute_batch("DELETE FROM relatedIdsToReupload")?;
+
+        Ok(())
+    }
+
+    /// Stores a snapshot of all locally changed items in a temporary table for
+    /// upload. This is called from within the merge transaction, to ensure that
+    /// changes made during the sync don't cause us to upload inconsistent
+    /// records.
+    ///
+    /// Conceptually, `itemsToUpload` is a transient "view" of locally changed
+    /// items. The local change counter is the persistent record of items that
+    /// we need to upload, so, if upload is interrupted or fails, we'll stage
+    /// the items again on the next sync.
+    fn stage_local_items_to_upload(&self) -> Result<()> {
+        // Stage remotely changed items with older local creation dates. These are
+        // tracked "weakly": if the upload is interrupted or fails, we won't
+        // reupload the record on the next sync.
+        self.db.execute_batch(
+            r#"
+            INSERT OR IGNORE INTO idsToWeaklyUpload(id)
+            SELECT b.id FROM moz_bookmarks b
+            JOIN mergedTree r ON r.mergedGuid = b.guid
+            JOIN moz_bookmarks_synced v ON v.guid = r.remoteGuid
+            WHERE r.useRemote AND
+                  /* "b.dateAdded" is in microseconds; "v.dateAdded" is in
+                     milliseconds. */
+                  b.dateAdded < v.dateAdded"#,
+        )?;
+
+        // Stage remaining locally changed items for upload.
+        self.db.execute_batch(&format!(
+            "
+            WITH RECURSIVE
+            {local_items_fragment}
+            INSERT INTO itemsToUpload(id, guid, syncChangeCounter, parentGuid,
+                                      parentTitle, dateAdded, title, placeId,
+                                      kind, url, keyword, position,
+                                      tagFolderName)
+            SELECT s.id, s.guid, s.syncChangeCounter, s.parentGuid,
+                   s.parentTitle, s.dateAdded, s.title, s.placeId,
+                   {kind}, h.url, NULL AS keyword, s.position,
+                   NULL AS tagFolderName
```

Comment:
Drop this: we can't have an outgoing `tagFolderName`, because we rewrite tag queries on the way in, and can't make new queries locally.